client/asset/btc/btc.go

```diff
@@ -766,7 +792,15 @@ func openSPVWallet(cfg *BTCCloneCFG) (*ExchangeWalletSPV, error) {
 
 	btc.node = loadSPVWallet(cfg.WalletCFG.DataDir, cfg.Logger.SubLogger("SPV"), peers, cfg.ChainParams)
 
-	return &ExchangeWalletSPV{btc}, nil
+	bday := time.Unix(int64(walletCfg.Birthday), 0)
```

Comment:
Here's part of what's fishy about forcing the birthday set in the `ConfigOpts`.  Here we may be preparing to start an **existing** `spvWallet` with a birthday already stored in the walletdb (not using `createSPVWallet`), but `NewWallet` > `openSPVWallet` will just get whatever birthday value and set it in `ExchangeWalletSPV.walletBirthday`.

Then later when the wallet is actually started with `connect`>`startSPVWallet`, it pulls it out of wallet.Manager's walletdb.  It could differ from what the caller gave to `NewWallet`.

However, I acknowledge that it looks good the way the reconfigure dialog works w.r.t. birthday:

![image](https://user-images.githubusercontent.com/9373513/151868720-d1f179f0-8aba-4077-85e5-65aceeb292e3.png)

I just find it a little goofy that changing wallet settings required "reminding" the wallet what it's birthday is when it's stored.

Sorry I have no concrete suggestions to improve this other than logging here and in createSPVwallet what birthday is being set

Reply:
In the beginning I had this birthday setting not a part of the wallet configuration, it was just another field in the `asset.CreateWalletParams` struct. We could put it back there. Then we could have the birthday be a parameter in the rescan request. On the UI, the rescan button would be on the wallet settings page as a separate section, and there would be a date input next to it. In this way there wouldn't be two dates in two separate databases. How do you feel about this?

Reply:
Hmm, that's an idea.  Since I commented I was imagining that a reconfigure with a new wallet birthday could detect a change an emit some kind of notification to suggest a rescan to have the new birthday take effect.

Let's see what @buck54321 thinks about that potential resolution to this UI/UX peculiarity since he requested doing this in `ConfigOpts` (for generally good reasons), although that was before your idea of putting the rescan button in a section of settings and tying birthday changes to a rescan.

I don't think the current solution is unacceptable, but it's somehow different from the other "settings", both in that it's awkward to constantly feed it back into wallet to change other settings, and in that it doesn't just "go into effect"

Reply:
I think this will be more natural for the user. There's really no point in changing the birthday if you're not going to do a rescan.

Reply:
@buck54321 Please let me know what's your opinion about this when you get a chance.

Reply:
If my read is right, we should pass in the `WalletConfig.Birthday` to `loadSPVWallet`, and then later check that in `(*spvWallet).startWallet` against the birthday stored in the btcwallet database. If there is a discrepancy, we would overwrite the btcwallet database value and (`if newBDay < oldBDay`) start a rescan. That's shouldn't be a problem, because we start a sync loop in `(*Core).connectWallet`, which is also called in `ReconfigureWallet`.

Reply:
Yeah, you're definitely on the same page.  My main hesitation with that was the last sentence in my previous comment on this, namely:

> We'd also have to worry about accidentally starting a wallet rescan with active swaps if it's triggered through ReconfigureWallet if the birthday appears to be changed.

I also still find it somewhat awkward that we're requiring a birthday when changing any other unrelated config change, but I think we can work around that by allowing a nil or zero-value birthday to indicate no change is intended.

Anyway, in this birthday-change-detected scenario, I'd sorta hope we could _not_ trigger an automatic rescan unless we tell the wallet that it is safe to do so.  Alternatively, possibly frontend could just advise to manually request a rescan seeing as it knows if the form was changed?

Reply:
> I also still find it somewhat awkward that we're requiring a birthday when changing any other unrelated config change, but I think we can work around that by allowing a nil or zero-value birthday to indicate no change is intended.

It'll be zero valued until the user specifies a custom birthday. Once the user specifies a custom birthday, the `WalletConfig.Settings` will always contain that birthday for us to compare, including during reconfiguration. So zero-value will be translated to default birthday by `ExchangeWallet`, but once a custom value is set it'll always be there. The "no change is intended" is signaled by equality of birthdays in `WalletConfig.Settings` and the btcwallet database.

> I'd sorta hope we could not trigger an automatic rescan unless we tell the wallet that it is safe to do so

> We'd also have to worry about accidentally starting a wallet rescan with active swaps if it's triggered through ReconfigureWallet if the birthday appears to be changed.

Does seem a little sketchy. Do you think that would end badly? If the risk is just a matter of some transient errors during the rescan, I'd be okay with it.

I'm with you 100% that wallet reconfiguration could use an overhaul, but I also don't have any solid suggestions for how to proceed. Maybe I'll open an issue. As it relates to this discussion, I guess a user shouldn't be allowed to set an earlier birthday if there are active orders because it requires a rescan? It's kind of odd though, because we already allow that the wallet may need to sync after a reconfigure, and we don't add any specific protections for the case of a long resync for other wallets. I guess we just assume that rpc wallet reconfigurations wouldn't result in a long sync? Any idea how long the rescan is with `PersistToDisk = true`?

I don't think a user should be able to set an earlier birthday without an automatic rescan.

Reply:
A normal sync up after reconfigure is really nothing compared to a rescan.  It could take an hour or more depending on peers and wallet history that might require fetching tons of actual blocks when filters match.  An automatic rescan would definitely be trouble.  Since the wallet's transaction history is dropped and rebuilt in the process, we really need to avoid initiating a rescan if there is anything active at all IMO.

Reply:
Hmm. Maybe you're right then that we should just skip the rescan and (somehow) send a notification for now, and work towards a better solution for reconfiguration later.

As a mid-term solution, we might consider a field in `ConfigOpt` like `DisableWhenActive bool` that would disable the input in the UI when the wallet has active orders.

Reply:
That `ConfigOption` could cover the UI.  Although the safest place to block a rescan would be in core when the new settings are actually submitted.  Perhaps `ReconfigureWallet` could inject `wcfg.Settings["special:permitAutoRescan"] = <bool>`.

Reply:
Ooh. I like that.

Reply:
> In the beginning I had this birthday setting not a part of the wallet configuration, it was just another field in the `asset.CreateWalletParams` struct. We could put it back there. Then we could have the birthday be a parameter in the rescan request. On the UI, the rescan button would be on the wallet settings page as a separate section, and there would be a date input next to it. In this way there wouldn't be two dates in two separate databases. How do you feel about this?

What about this idea? It seems like it would solve all the complications. Just block rescans when there is anything active happening. It would be much more obvious for the user that once they click the rescan button, they will not be able to use the wallet for a while.

Reply:
Maybe. How does the user see what the currently set birthday is? 

I'm hesitant to add custom fields to `CreateWalletParams` that are specific to an asset. The `ConfigOpts`/`Settings` system was created to reign in the asset-specific configuration and keep the API sensible. 

I'm sort of on the fence here. We **should** just overhaul the wallet configuration system. @martonp why don't you choose whatever solution seems like the least work, and then we can bang our heads together on a new configuration system to squeeze into Phase III. I personally am leaning towards `ConfigOpts.DisableWhenActive` **+** `wcfg.Settings["special:permitAutoRescan"] = <bool>`, but just barely.

Reply:
> Maybe. How does the user see what the currently set birthday is?

That's a good question, there's not really a mechanism to send it at the moment. I guess ideally there would be creation only settings, and then maybe some kind of way to retrieve some read only information from the wallet. I'll just go with `ConfigOpts.DisableWhenActive` + `wcfg.Settings["special:permitAutoRescan"] = <bool>`.